### PR TITLE
DEP: Deprecate passing a DataFrame to from_records

### LINKED ISSUE #51353: PERF: DataFrame.from_records with DataFrame input
Speculative: DataFrame.from_records if passed a DataFrame goes through internals.construction.to_arrays and ends up splitting the frame into individual columns.  This can result in slower subsequent operations if that splitting turned out to be unnecessary.  Can we handle DataFrame input inside from_records and avoid this splitting?

This is also the only way we get to to_arrays with a DataFrame input, so the function could be simplified if that case could be ruled out.

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #51353 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

cc @jbrockmendel  as discussed, restructuring this allows us to simplify to_arrays immediately

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -2171,6 +2174,23 @@ def from_records(
         2      1     c
         3      0     d
         """
+        if isinstance(data, DataFrame):
+            warnings.warn(
+                "Passing a DataFrame to DataFrame.from_records is deprecated. Use "
+                "set_index and/or drop instead to modify the DataFrame.",
```
Comment (jbrockmendel) [inline pandas/core/frame.py:2180]: "instead"?

Comment (phofl) [reply pandas/core/frame.py:2180]: Why I am using instead or placement of instead within the sentence?

**Comment by jbrockmendel** [comment]
small comment, otherwise LGTM

File: pandas/core/frame.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/frame.py:2180]: im suggesting adding "instead" at the end of the last sentence

Comment (phofl) [reply pandas/core/frame.py]: changed

**Pull request merged**